core-bundle/src/Filesystem/Dbafs.php

```diff
@@ -0,0 +1,863 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Filesystem;
+
+use Contao\CoreBundle\Event\ContaoCoreEvents;
+use Contao\CoreBundle\Event\RetrieveDbafsMetadataEvent;
+use Contao\CoreBundle\Event\StoreDbafsMetadataEvent;
+use Doctrine\DBAL\Connection;
+use League\Flysystem\FileAttributes;
+use League\Flysystem\FilesystemAdapter;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+use Symfony\Component\Filesystem\Path;
+use Symfony\Component\Uid\Uuid;
+use Symfony\Contracts\Service\ResetInterface;
+
+/**
+ * @phpstan-import-type Record from DbafsInterface
+ * @phpstan-import-type ExtraMetadata from DbafsInterface
+ */
+class Dbafs implements ResetInterface, DbafsInterface
+{
+    public const FILE_MARKER_EXCLUDED = '.nosync';
+    public const FILE_MARKER_PUBLIC = '.public';
+
+    private Connection $connection;
+    private EventDispatcherInterface $eventDispatcher;
+
+    private string $table;
+    private string $hashAlgorithm;
+    private string $dbPathPrefix = '';
+    private int $maxFileSize = 2147483648; // 2 GiB
+    private int $bulkInsertSize = 100;
+
+    /**
+     * @var array<string, array|null>
+     * @phpstan-var array<string, Record|null>
+     */
+    private array $records = [];
+
+    /**
+     * @var array<string, string|null>
+     */
+    private array $pathByUuid = [];
+
+    /**
+     * @var array<int, string|null>
+     */
+    private array $pathById = [];
+
+    public function __construct(Connection $connection, EventDispatcherInterface $eventDispatcher, string $table, string $hashAlgorithm)
```

Comment:
Proposal: Let's make the hash algorithm its own service, it would provide multiple advantages:

```php
interface FileContentHasherInterface
{
    public function getHash(string $fileContents): string
}
```

The default service could work like this:

```php
class NativeFileContentHasher implements FileHasherInterface
{
    private ?string $nativeFunctionName = null;

    public function __construct(string $nativeFunctionName)
    {
        $this->nativeFunctionName = $nativeFunctionName;
    }

    public function getHash(string $fileContents): string
    {
        return hash($this->nativeFunctionName, $fileContents);
    }

    public static function createWithBestSuitedFromPlatform(): self
    {
        $algo = 'md5';

        if (PHP_VERSION_ID > 80100) {
            $algo = 'xxh128';
        }

        return new self($algo);
    }
}
```

Advantages:

- Easy to reuse across all `dbafs` configurations because they'll most likely use the same algorithm so they can all reuse the same service.
- Automatically uses the best native hash function available by default without us having to do anything (just use the factory method as service factory)
- Easy to test custom alternatives and see how they perform
- Better testability

Things to consider:

- What happens if the algorithm changes? No hash will match anymore, does that mean, we need a migration or can we do something like the password API (`password_*`) does internally. E.g. storing `$md5$<hash>` or `$xxh128$<hash>` and some `needsRehash()`? Basically really the same as the password API just not with cryptographically secure hashes?
- Just a wild thought: Might also be interesting to pass on the operator to that interface? I'm just thinking that maybe it's not needed to request a 4 GB file from an S3 storage in order to calculate it's hash if the operator already knows this and it's just metadata on the file. But then again, it's probably mixing different hash algorithms because S3 uses md5 and someone else something different. But there's always only one operator per file path so...could work ;)

Reply:
> What happens if the algorithm changes? No hash will match anymore, does that mean, we need a migration 

Changing the algorithm and directly triggering a sync afterwards (i.e. not altering the filesystem) should rebuild the table just fine - it should be analogous to changing the contents of all files at once but still keeping UUIDs/paths the same.

It should IMHO be a user decision for this reason, though. Making it dependent of the `PHP_VERSION_ID` will probably cause confusion when people switch environments/deploy stuff. I think I would change the default for Contao 5 where we have PHP8 as a minimum requirement anyways.


> Let's make the hash algorithm its own service,

Currently you can simply change a value in the config (this is how I tested switching algorithms) to test out and set different algorithms. I mean, I'm all for abstracting stuff but PHP's `hash(string $algo, string $data)` already kind of does that, no? :slightly_smiling_face:

Reply:
Yeah but the hash algo is part of every `dbafs` config so you have to make sure you update all of them. But yeah, it might be true that it's unnecessary. Just wondering if that whole concept is gonna work out with remote FS. I mean...calculating a hash for a 4GB remote file is going to take a while. That why I'm thinking about maybe somehow outsourcing that to the adapter. Because, well...is `xxh128` faster than `md5`? Hell yes. But is it also faster with S3 if in fact the `md5` hash was part of the meta data and nothing needs to be downloaded? Hell no.

Reply:
Yes, I think the algorithm choice is more something for the (fast) local filesystem which will be the default for 99% of our users. And it's probably more about the default and the ability to configure it _should_ you want to tweak it, rather than a setting everyone has to fiddle with.

Your idea from https://github.com/contao/contao/pull/3774#discussion_r767493300 will probably make the larger difference when using a DBAFS with slow filesystems. If you want to benefit from filesystems that can provide file hashes via their metadata, I've got the feeling you will end up writing your own specialized `DbafsInterface` anyways.

Reply:
I thought about this again and I think it's really a major design flaw to always execute the file content hashing locally. It **really should** happen in the adapter because things can be easily precalculated on any remote host instead of transferring thousands of megabytes only for a hash. However, I did some research on Flysystem and why this is not a first class citizen like e.g. `lastModified()`. And it seems like it's mostly because it's just not standardized. Not even in S3. But can I convince you to at least have something like `$hash = $event->getHash() ?: hash(....);` or so?

Reply:
An abstraction for the hashing is now in place with 7b97f8cf41d13bfcd8f14290b91a2f6bde97d3e5 and using "last modified" to skip generating hashes is catered for in bf7c0b07aa61983aa05d5b1fb4f6afa837467e3f (not implemented in the `Dbafs`, yet, but part of the interface).